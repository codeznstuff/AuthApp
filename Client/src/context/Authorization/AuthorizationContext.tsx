import React, { createContext, useContext } from 'react';
import { useAsync } from 'react-async';
import { fetchUser, IApplicationUser } from '../../services/AuthorizationApi';
import { getMsal } from '../../utils/authentication';
import { KiewitSpinner } from '../../components/common/KiewitSpinner/KiewitSpinner';
import UnauthenticatedApp from '../../UnauthenticatedApp';

export const AuthorizationContext = createContext<any>({
  currentUser: { UserId: 0 }
});

const loadData = async (): Promise<IApplicationUser | Error> => {
  try {
    // Make API call for MSAL AAD Authentication
    //const authUser = await getMsal();
    // Make API call to get user data from Authorization API
    //const user = await fetchUser(authUser.accountIdentifier);
    const user = await fetchUser('be64b27b-59b6-43bb-a824-51c16d328ffb');
    return user;
  } catch (error) {
    const ex: Error = error;
    return ex;
  }
};

export const AuthorizationProvider = (props: any): any => {
  const { data, isPending } = useAsync({
    promiseFn: loadData
  });

  if (isPending) {
    return <KiewitSpinner />;
  }
  if (data instanceof Error) {
    return <UnauthenticatedApp />;
  }
  return <AuthorizationContext.Provider value={data} {...props} />;
};

export const useAuthorization = (): IApplicationUser => {
  const context: any = useContext(AuthorizationContext);
  if (context === undefined) {
    throw new Error(`useAuthorization must be used within an AuthorizationProvider`);
  }
  return context;
};
