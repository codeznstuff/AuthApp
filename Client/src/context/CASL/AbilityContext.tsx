import React, { createContext, useContext } from 'react';
import { createContextualCan } from '@casl/react';
import { AbilityBuilder } from '@casl/ability';
import { useAuthorization } from '../Authorization/AuthorizationContext';
import { IClaimData, IApplicationUser } from '../../services/AuthorizationApi';

// Create Ability Context
export const AbilityContext = createContext(null);

// Create Can context that uses Ability Context
export const Can = createContextualCan(AbilityContext.Consumer);

// Creates the Ability Context Provider
export const AbilityProvider = (props: any): any => {
  const user = useAuthorization();
  const ability = defineRules(user);
  return <AbilityContext.Provider value={ability} {...props} />;
};

// Returns the type for the object passed into Can
const subjectName = (subject: any) => {
  if (!subject || typeof subject === 'string') {
    return subject;
  }
  if (subject.hasOwnProperty('userId')) {
    return 'user';
  }
  return typeof subject;
};

// Establishes the Permissions for the application
const defineRules = (user: IApplicationUser) => {
  return AbilityBuilder.define({ subjectName }, (can: any) => {
    // FOR DEMO ONLY!!!
    let adminClaim: any;
    var min = 0;
    var max = 1;
    var random = Math.floor(Math.random() * (max - min + 1) + min);
    if (random === 1) {
      adminClaim = user.claims ? user.claims.filter((claim: IClaimData) => claim.claimType === 'Role' && claim.claimValue === 'Admin') : [];
    } else {
      adminClaim = [];
    }

    // Admin roles gives full site access
    // const adminClaim = user.claims ? user.claims.filter((claim: IClaimData) => claim.claimType === 'Role' && claim.claimValue === 'Admin') : [];
    if (adminClaim.length > 0) {
      can(['read', 'create', 'edit', 'delete'], 'user');
      can('manage', 'all');
    }
    // Standard user permissions
    else {
      can(['edit', 'delete'], 'user', { userId: user.userId });
      can(['read', 'create'], 'user');
    }
  });
};

// Context Hook
export const useAbility = (): any => {
  const context: any = useContext(AbilityContext);
  if (context === undefined) {
    throw new Error(`useAbility must be used within an AbilityProvider`);
  }
  return context;
};
