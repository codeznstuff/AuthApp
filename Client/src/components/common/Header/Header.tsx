import React from 'react';
import { Link } from 'react-router-dom';
import { Navbar, Nav } from 'react-bootstrap';
import { useAuthorization } from '../../../context/Authorization/AuthorizationContext';
import { useAbility } from '../../../context/CASL/AbilityContext';
import PageAlerts from '../PageAlerts/PageAlerts';

export const Header = () => {
  const user = useAuthorization();
  const ability = useAbility();
  const isAdmin = ability.can('manage', 'all');

  return (
    <>
      <Navbar bg="light" variant="light" expand="sm">
        <Link className="navbar-brand" to="/">
          Authorization Demo
        </Link>
        <Navbar.Toggle aria-controls="navbar-nav" />
        <Navbar.Collapse id="navbar-nav">
          <Nav className="mr-auto">
            <Link className="nav-link" to="/">
              All Users
            </Link>
            {isAdmin && <Nav.Link href="/">Admin Only</Nav.Link>}
          </Nav>
          <Navbar.Text>
            {user.firstName} {user.lastName}
          </Navbar.Text>
        </Navbar.Collapse>
      </Navbar>
      <PageAlerts />
    </>
  );
};
