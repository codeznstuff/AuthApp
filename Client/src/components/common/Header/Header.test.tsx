import React from 'react';
import { render, RenderResult, cleanup } from '@testing-library/react';
import { Header } from './Header';
import { AuthorizationContext } from '../../../context/Authorization/AuthorizationContext';
import { BrowserRouter } from 'react-router-dom';
import { AbilityContext } from '../../../context/CASL/AbilityContext';
import { AbilityBuilder } from '@casl/ability';

describe('Header', () => {
  let header: RenderResult;
  afterEach(cleanup);

  describe('when user is not an account admin', () => {
    beforeEach(() => {
      const data: any = { user: { userId: 0 } };
      const ability = AbilityBuilder.define((can: any) => {
        can(['read'], 'user');
      });
      header = render(
        <BrowserRouter>
          <AuthorizationContext.Provider value={data}>
            <AbilityContext.Provider value={ability}>
              <Header />
            </AbilityContext.Provider>
          </AuthorizationContext.Provider>
        </BrowserRouter>
      );
    });

    it('matches snapshot', () => {
      expect(header.container.firstChild).toMatchSnapshot();
    });
  });

  describe('when user is an account admin', () => {
    beforeEach(() => {
      const data: any = { user: { userId: 0 } };
      const ability = AbilityBuilder.define((can: any) => {
        can(['delete'], 'user');
      });
      header = render(
        <BrowserRouter>
          <AuthorizationContext.Provider value={data}>
            <AbilityContext.Provider value={ability}>
              <Header />
            </AbilityContext.Provider>
          </AuthorizationContext.Provider>
        </BrowserRouter>
      );
    });

    it('matches snapshot', () => {
      expect(header.container.firstChild).toMatchSnapshot();
    });
  });
});
