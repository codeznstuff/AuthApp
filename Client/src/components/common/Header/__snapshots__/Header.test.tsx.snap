// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header when user is an account admin matches snapshot 1`] = `
<nav
  class="navbar navbar-expand-sm navbar-light bg-light"
>
  <a
    class="navbar-brand"
    href="/"
  >
    Authorization Demo
  </a>
  <button
    aria-controls="navbar-nav"
    aria-label="Toggle navigation"
    class="navbar-toggler"
    type="button"
  >
    <span
      class="navbar-toggler-icon"
    />
  </button>
  <div
    class="navbar-collapse collapse"
    id="navbar-nav"
  >
    <div
      class="mr-auto navbar-nav"
    >
      <a
        class="nav-link"
        href="/"
      >
        All Users
      </a>
    </div>
    <span
      class="navbar-text"
    >
       
    </span>
  </div>
</nav>
`;

exports[`Header when user is not an account admin matches snapshot 1`] = `
<nav
  class="navbar navbar-expand-sm navbar-light bg-light"
>
  <a
    class="navbar-brand"
    href="/"
  >
    Authorization Demo
  </a>
  <button
    aria-controls="navbar-nav"
    aria-label="Toggle navigation"
    class="navbar-toggler"
    type="button"
  >
    <span
      class="navbar-toggler-icon"
    />
  </button>
  <div
    class="navbar-collapse collapse"
    id="navbar-nav"
  >
    <div
      class="mr-auto navbar-nav"
    >
      <a
        class="nav-link"
        href="/"
      >
        All Users
      </a>
    </div>
    <span
      class="navbar-text"
    >
       
    </span>
  </div>
</nav>
`;
